spacy/cli/project/run.py

```diff
@@ -93,13 +119,30 @@ def project_run(
         check_spacy_commit = check_bool_env_var(ENV_VARS.PROJECT_USE_GIT_VERSION)
         with working_dir(project_dir) as current_dir:
             msg.divider(subcommand)
-            rerun = check_rerun(current_dir, cmd, check_spacy_commit=check_spacy_commit)
+            rerun = check_rerun(
+                current_dir,
+                cmd,
+                check_spacy_commit=check_spacy_commit,
+                mult_group_mutex=mult_group_mutex,
+            )
             if not rerun and not force:
                 msg.info(f"Skipping '{cmd['name']}': nothing changed")
             else:
                 run_commands(cmd["script"], dry=dry, capture=capture)
                 if not dry:
-                    update_lockfile(current_dir, cmd)
+                    update_lockfile(current_dir, cmd, mult_group_mutex=mult_group_mutex)
+
+
+def _get_workflow_steps(workflow_items: List[Union[str, List[str]]]) -> List[str]:
+    steps: List[str] = []
+    for workflow_item in workflow_items:
+        if isinstance(workflow_item, str):
+            steps.append(workflow_item)
+        else:
+            assert isinstance(workflow_item, list)
+            assert isinstance(workflow_item[0], str)
+            steps.extend(workflow_item)
+    return steps
 
 
```

Comment:
In these two cases (user does not enter a command or asks for `--help` about a command) I have abstracted away the parallel execution and just listed the commands in the workflow. In the output of `spacy project document`, on the other hand, I have made the parallel execution explicit. I think this is a good compromise between making everything too complicated and providing the necessary information; IMHO if someone just wants to know what a command does they are probably not that interested in implementation details like serial or parallel execution. I am open to being persuaded otherwise though :-).

Reply:
I personally use `spacy project run` like ALL THE TIME when working with a spacy project, and I'd personally like to see the parallellization represented there. Would it be difficult to implement?

Reply:
There would be no difficulty at all in implementing it. The question is rather how it could be represented in a way that informs rather than confuses the user who may not necessarily be interested in such technical details as parallel vs. serial execution. Do you have a suggestion for _how_ you would like to see this information shown?

Reply:
Right now it would print
> all   setup -> commandA -> commandB

for me personally, something like

> all   setup -> commandA
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-> commandB

(with the `->` aligned) would make it more clear that this is not a sequence but a parallel execution. But I'm not sure how feasible it is to actually implement that and make sure the vertical alignment works across different terminals etc. 

Alternatively, it could be something like 
> all   setup -> [commandA, commandB]

which would also make it easier to chain with subsequent steps.

Reply:
That is a clear syntax for the case where somebody types `spacy project run`. However, what I was really unsure how to tackle was the following situation:

```
> spacy project run all --help
Usage: python -m spacy project run all 

Workflow consisting of 3 commands:

1. setup      
2. commandA   
3. commandB   
```

Perhaps this syntax:

```
> spacy project run all --help
Usage: python -m spacy project run all 

Workflow consisting of 3 commands:

1.  setup      
2=. commandA   
2=. commandB   
```
?

Reply:
Could we list them as
```
1. setup      
2a. commandA   
2b. commandB   
```
?

Reply:
Another option would be to add a single-character column to the table to the right of the numbers and use Ascii art to bracket together the multiprocessing groups (see `visualisation.py` [here](https://github.com/explosion/spaCy/pull/9930/files#diff-beda9f8c93177c5d5367ce855433a713f8ab9f70f781ec4cf1826aa46f52e30f) for an example of the technique I mean)

Reply:
I have an idea on syntax, but before I get there, I have a question about the `project.yml` syntax for this: is the only way to designate parallelism by nesting commands under `workflows`? I think that could be confusing to a user who wants to nest commands for a workflow but _doesn't_ intend for them to be run in parallel. Nesting would be a nice feature on it's own for organizing more complex project workflows.

Building off of that, if we first were to incorporate serial nested workflows, you could use some sort of syntax in those nested workflows to indicate parallelism - something like surrounding the commands with `(` and `)`. So for example, this would still be a serial project workflow:

```yaml
workflows:
  all:
    - setup
    - nestedCommandGroup
       - commandA
       - commandB
```

And would be the same as executing without the nested group while providing the user a way to indicate `commandA` and `commandB` are related.

Then, if you wanted to indicate those could be run in parallel, you could have something like:

```yaml
workflows:
  all:
    - setup
    - nestedCommandGroup
       - (commandA)
       - (commandB)
```

Then for `spacy project run` I'd probably prefer a hybrid of what's already been proposed, specifically:

```
1.  setup      
2. (commandA)
2. (commandB)
```

This way the indicator of parallelism lies with the command name, _not_ the numbering—which feels like we might be overloading it with this information.

Reply:
The big problem here is that `(commandA)` and `(commandB)` are valid command identifiers in the pre-existing system. This means that such identifiers may already be present in config files that people are using in a non-nested context, so that the only feature reliably indicating parallelism in the above `spacy project run` output would be the numbering.

It would be theoretically possible to implement a rule where bracketed names denoted parallelism in a nested context but were processed verbatim in a non-nested context, but this would get very confusing in config files with both types of bracketing, be complex to explain and to code and would also require rules to deal with cases like 

```
workflows:
  all:
    - setup
    - nestedCommandGroup
       - (commandA)
       - commandB
       - (commandC)
```

Because of the backwards-compatibility problem, it's probably best to avoid any solution based around the command name.

I think generally that nesting is as good a way as any to denote parallelism. It was Matt's original suggestion in the Ora ticket, although I don't know to what extent it was a considered suggestion and to what extent it was just an example. But I think it's going to get confusing for users if we introduce nesting as a supported option at the same time as introducing parallelism but without a strict relationship between the two. And if users just want to group commands they can do so using spacing between the command names.

Reply:
**Logging**: from the suggested options I like the ASCII art one best. What about having both a command ID and a _timestep_ ID? Like, `setup` would be 1/1, `commandA` would be 2/2, `commandB` would be 3/2. Perhaps splitting the logical step from the time step would clear things up.
```
Workflow consisting of 3 commands:
| Command ID | Time step | Command  |
|------------|-----------|----------|
| 1          | 1         | setup    |
| 2          | 2         | commandA |
| 3          | 2         | commandB |
```

**Syntax in project.yml**: Like the nesting idea, but agree with @richardpaulhudson on potential confusion and the downsides of using brackets to denote parallelism. How about we make this more explicit on the level of the name of the nested group? For example:
```yaml
workflows:
  all:
    - setup
    - nestedCommandGroup::parallel
       - commandA
       - commandB
```
This would indicate that all commands in the nested group would be executed in parallel (I'd go for all or nothing in one nested group, I think mixing would complicate things too much).
Having a nested command group without `::parallel` would imply serial execution.

The downside to this would be adding an arbitrary marker like `::parallel`. Alternatively we could handle in proper  .yaml-style with dicts:
```yaml
workflows:
  all:
    - setup
    - nestedCommandGroup
        mode: parallel
        commands:
          - commandA
          - commandB
```

Both approaches are backwards-compatible, since these suggestions only affect nested command groups.

Reply:
This looks good. However, I'm finding it hard to see that people wouldn't be confused by nested command groups that didn't actually have any additional semantics, and wonder whether it shouldn't be a general principle of config files that different syntax corresponds to different semantics?

Reply:
I'd enforce the same syntax for command groups, regardless of whether they are intended to be executed in parallel or not. For nested sequential command groups:
```yaml
workflows:
  all:
    - setup
    - nestedCommandGroup
        mode: sequential
        commands:
          - commandA
          - commandB
```
Or, if we have `mode` default to `sequential`:
```yaml
workflows:
  all:
    - setup
    - nestedCommandGroup
        commands:
          - commandA
          - commandB
```
Does this address your point?

Reply:
That much was clear, but I'm troubled by the fact that (unless I've misunderstood something):

```
workflows:
  all:
    - setup
    - nestedCommandGroup
        commands:
          - commandA
          - commandB
```

 would have identical semantics to
 
```
workflows:
  all:
    - setup
    - commandA
    - commandB
```

Generally in a config file I see it as being undesirable for there to be multiple ways of achieving the same behaviour.

Reply:
Ah, I see - misunderstood you before. Yes, I agree. Unless nested command groups have some other user-modifiable properties (should they? Not in the loop with this one), it doesn't make too much sense to have them in a sequential version IMO.

Reply:
Yeah, I agree that's potentially confusing to have the multiple behaviors for the same end result, but it's also a totally optional change from the flat to nested syntax. I think this pattern of short "string" vs. long "dict" values is pretty common - e.g. [`docker-compose` ports: Short and Long syntax](https://docs.docker.com/compose/compose-file/compose-file-v3/#ports). To me the implication that a nested group is automatically parallel is more confusing because the user never explicitly states they want parallelism. It also locks us into that behavior for any nesting of commands, having to make _future_ changes to nesting backwards-compatible with the parallelism—which is why I proposed distinguishing serial from parallel in some way other than the implication from nesting.

I have an alternative idea from the closest thing I've used to the spaCy project config, which is [Taskfile](https://taskfile.dev/usage/#task-dependencies). The model there is a little different because you can set dependent tasks, and if a task has multiple dependencies, they'll automatically be run in parallel. They have an option to run specific non-dependent commands in parallel when they're executed, so translating to our use case: rather than defining parallelism in the flag, we could make `spacy project run` take a `--parallel` arg and a variable number of args that are commands to run in parallel. 

The main downside with that I see is that there's no way to document that commands _can/should_ be run in parallel in the config. However, part of me thinks that parallelism semantically makes more sense as an execution option (with the `run` command), rather than something that is in the config, mostly because I can see it being easy to modify a command in your config and have it silently fail because it's now dependent on another command.

Reply:
I can see the complaint that the 'unnamed nesting' thing is too cryptic and the parallelism will be surprising. But I think the naming feels surprising, it doesn't do anything if it's done serially like that and it feels too overlapping with the subcommands as has been said.

I think I'd like this better than the named subgrounds with an option 'parallel':

```
workflows:
  all:
    - setup
    - parallel:
      - commandA
      - commandB
```

In terms of the grammar of the config it feels a little ugly to me that you can have this object that has one key, (`parallel`), having a list there feels a little more aesthetic to me. But the object would maybe make things a bit more flexible in future, and does make it more obvious what's going on.